source/agora/common/Ensure.d

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+
+     Throw a statically allocated exception with formatting capacility.
+
+     The function in this module are not `@nogc` / `pure` but they do not
+     allocate if the formatting does not allocate.
+     A static buffer is used by the Exception, so the output length is limited.
+     Do not try to print a full block, for example.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Ensure;
+
+import ocean.text.convert.Formatter;
+
+/// Module ctor to initialize the thread-local instance
+static this ()
+{
+    instance = new FormattedException();
+}
+
+/// The underlying instance
+private FormattedException instance;
+
+/// An exception which doesn't allocate but allows formatting a message
+private final class FormattedException : Exception
+{
+    /// Internal buffer, should be enough for most messages
+    private char[2048] buffer;
+    /// Used size
+    private size_t end;
+
+    private this () @safe pure nothrow @nogc
+    {
+        super("An Exception occured");
+    }
+
+    public override const(char)[] message () const return
+        @safe pure nothrow @nogc
+    {
+        return this.buffer[0 .. this.end];
+    }
+}
+
+public void ensure (Args...) (bool exp, string fmt, lazy Args args)
+{
+    if (!exp)
+    {
+        auto res = instance.buffer.snformat(fmt, args);
+        instance.end = res.length;
```

Comment:
Maybe here can use min of `res.length` and `2048` ?

Reply:
Not needed as explained above `snformat` will truncate.

Reply:
`snformat` will never return more than the buffer size.
[Here is the source of `snformat`](https://github.com/sociomantic-tsunami/ocean/blob/adb31c84baa2061d07aaa0cb7a7d14c3cc98309b/src/ocean/text/convert/Formatter.d#L200-L215)

Annotated below:
```D
public mstring snformat (Args...) (mstring buffer, cstring fmt, Args args)
{
    size_t start; // Used to keep track of where we are in `buffer`

    // The sink passed to `sformat`, same technique as our serialization library
    scope FormatterSink sink = (cstring s)
    {
        // 0 <= left < buffer.length. If we have filled the buffer, `left` will be 0.
        size_t left = buffer.length - start;
        // Likewise, 0 <= wsize <= s.length
        // If left is 0, `wsize` will be 0 (wsize stands for "write size")
        size_t wsize = left <= s.length ? left : s.length;
        if (wsize > 0)
            buffer[start .. start + wsize] = s[0 .. wsize];
        start += wsize; // If the buffer is full, it means start must be == buffer.length, and we just add 0 to it
    };

    sformat(sink, fmt, args);
     // Here you can see that `min` is not needed since we would get a bounds check error if `start > buffer.length`
    return buffer[0 .. start];
}
```